### ISSUE TITLE
发现 bn 没有和 torch 对齐

### ISSUE THREAD

**Issue by ccssu** [Oneflow-Inc/oneflow#8913]
## Summary

最近在用 oneflow 修改 torch版本yolov5代码，出现loss无法对齐，发现 oneflow bn 没有 和 torch 对齐。
## Code to reproduce bug

```python
import numpy as np
in_channels = 1
out_channels = 1
batch_size = 1

input_arr = [batch_size,in_channels, 4, 4]

def torch_cal(np_arr: np.number)-> np.number:
    import torch
    torch_imgs = torch.FloatTensor(np_arr)
    # print(torch_imgs.dtype)
    torch_bn = torch.nn.BatchNorm2d(1)
    torch_y = torch_bn(torch_imgs)
    return torch_y.cpu().detach().numpy().flatten()

def oneflow_cal(np_arr: np.number)-> np.number:
    import oneflow as torch
    torch_imgs = torch.FloatTensor(np_arr)
    # print(torch_imgs.dtype)
    torch_bn = torch.nn.BatchNorm2d(1)
    torch_y = torch_bn(torch_imgs)
    return torch_y.cpu().detach().numpy().flatten()

def test_00():
    test_np_arr = [0.54384547,0.49331503,0.82029225,0.29535902,0.37179976,0.1883129,
 0.72609789, 0.14132864, 0.86247542, 0.8728756,  0.38950897, 0.65364231,
 0.45458509, 0.16700907, 0.6270173,  0.28827853]
    np_arr = np.array(test_np_arr)
    np_arr.resize(input_arr)
    a = torch_cal(np_arr)
    b = oneflow_cal(np_arr)
    print(np.allclose(a, b))

def test_01(len : int):
    for i in range(len):
        np_arr = np.random.random(input_arr)
        a = torch_cal(np_arr)
        b = oneflow_cal(np_arr)
        print(np.allclose(a,b))
        if np.allclose(a,b) == False:
            print(a.flatten().tolist())
            print(b.flatten().tolist())

# 一个固定的numpy测试
test_00()

# 随机生成 20 numpy测试 
test_01(20)
```
```output
False
True
True
True
True
True
True
True
True
False
[1.984910488128662, -0.01789187639951706, 0.5719995498657227, -1.9109092950820923, -0.6650400757789612, 1.1599366664886475, -0.0525544136762619, 0.31155598163604736, 0.36744606494903564, -1.4633088111877441, -0.5016557574272156, 0.2445223480463028, -1.4118359088897705, 0.8218916654586792, 0.9027228355407715, -0.3417889475822449]
[1.984911561012268, -0.017892157658934593, 0.5719996690750122, -1.910910725593567, -0.6650407910346985, 1.1599371433258057, -0.052554719150066376, 0.3115559220314026, 0.36744606494903564, -1.4633100032806396, -0.5016563534736633, 0.24452222883701324, -1.411837100982666, 0.8218919038772583, 0.9027231931686401, -0.3417894244194031]
True
True
False
[1.0973858833312988, -1.4813170433044434, -0.17858001589775085, -1.6678498983383179, -0.3279232978820801, -0.6441431045532227, 1.3837538957595825, 1.329971432685852, 0.00395537493750453, -0.5098833441734314, 1.5887246131896973, 0.5017133355140686, 0.11579941213130951, -0.9867441654205322, 0.762543261051178, -0.9874081611633301]
[1.0973867177963257, -1.481318473815918, -0.1785803586244583, -1.6678515672683716, -0.3279237747192383, -0.6441438794136047, 1.3837549686431885, 1.3299723863601685, 0.003955202642828226, -0.5098839402198792, 1.5887258052825928, 0.5017135739326477, 0.11579933762550354, -0.9867452383041382, 0.762543797492981, -0.987409234046936]
True
True
False
[-1.2112410068511963, 1.7817611694335938, 0.6818868517875671, 0.955621600151062, -0.007169682066887617, 1.1772127151489258, 0.008479944430291653, -1.555105447769165, 0.5366457104682922, -0.3083673119544983, -0.9230159521102905, 0.3319283723831177, 1.1681081056594849, -0.09982077777385712, -0.9252980351448059, -1.6116262674331665]
[-1.2112408876419067, 1.7817612886428833, 0.6818869709968567, 0.9556216597557068, -0.007169574499130249, 1.1772128343582153, 0.008480051532387733, -1.555105447769165, 0.5366458296775818, -0.30836719274520874, -0.923015832901001, 0.3319284915924072, 1.1681082248687744, -0.09982067346572876, -0.9252979755401611, -1.611626148223877]
True
True
True
True
True
```
## System Information
```
>>> import torch
>>> oneflow.__version__
'0.8.1.dev20220804+cu112'
>>> torch.__version__
'1.11.0+cu102'
```

**Comment by ShiyuanResearch** [comment]
a-b的误差有多大，上周也有类似的issue

**Comment by BBuf** [comment]
cpu确实有误差，需要看一下cuda上是否正常。

**Comment by ccssu** [comment]
看看这份代码 应该是cuda上的
```python
import numpy as np
import time

in_channels = 3
out_channels = 32
batch_size = 16

input_arr =  [16, 3, 640, 640]
# weight_arr = [32, 3, 6, 6]
def torch_cal(np_arr: np.number, weight_up = None)-> np.number:
    import torch
    torch_imgs = torch.FloatTensor(np_arr).cuda()
    # print(torch_imgs.dtype)
    # torch_m = torch.nn.Conv2d(in_channels, out_channels, 6, 6, 2, 2, bias=False)
    # torch_m.weight = torch.nn.Parameter(torch.tensor(weight_up, dtype=torch.float32))
    # torch_m.to("cuda")
    torch_bn = torch.nn.BatchNorm2d(3).cuda()
    torch_y = torch_bn(torch_imgs)
    return torch_y.cpu().detach().numpy()

def oneflow_cal(np_arr: np.number, weight_up = None)-> np.number:
    import oneflow as torch
    torch_imgs = torch.FloatTensor(np_arr).cuda()
    # print(torch_imgs.dtype)
    # torch_m = torch.nn.Conv2d(in_channels, out_channels, 6, 6, 2, 2, bias=False)
    # torch_m.weight = torch.nn.Parameter(torch.tensor(weight_up, dtype=torch.float32))
    # torch_m.to("cuda")
    torch_bn = torch.nn.BatchNorm2d(3).cuda()
    torch_y = torch_bn(torch_imgs)
    return torch_y.cpu().detach().numpy()


def test_01(len : int):
    for i in range(10):
        np_arr = np.random.random(input_arr)
        a = torch_cal(np_arr)
        b = oneflow_cal(np_arr)
        res = np.allclose(a, b)
        print(res)
        if res == False:
            print(a.flatten().tolist()[0:15])
            print(b.flatten().tolist()[0:15])
            break
           



# 随机生成 20 numpy测试 
test_01(20)
```

```output
False
[1.6659996509552002, -0.7920742034912109, 1.3852777481079102, -0.3436308205127716, 1.0656812191009521, -1.3124674558639526, -1.5769178867340088, 1.3032417297363281, -0.16378989815711975, 0.3524649739265442, -0.8447245359420776, 1.1981629133224487, -1.5700066089630127, 0.4203341007232666, 0.928550124168396]
[1.6659995317459106, -0.7920739650726318, 1.3852777481079102, -0.34363067150115967, 1.0656812191009521, -1.3124672174453735, -1.5769174098968506, 1.3032417297363281, -0.163789764046669, 0.35246503353118896, -0.8447242975234985, 1.1981627941131592, -1.570006251335144, 0.420334130525589, 0.928550124168396]
```

**Comment by BBuf** [comment]
> 看看这份代码 应该是cuda上的
> 
> ```python
> import numpy as np
> import time
> 
> in_channels = 3
> out_channels = 32
> batch_size = 16
> 
> input_arr =  [16, 3, 640, 640]
> # weight_arr = [32, 3, 6, 6]
> def torch_cal(np_arr: np.number, weight_up = None)-> np.number:
>     import torch
>     torch_imgs = torch.FloatTensor(np_arr).cuda()
>     # print(torch_imgs.dtype)
>     # torch_m = torch.nn.Conv2d(in_channels, out_channels, 6, 6, 2, 2, bias=False)
>     # torch_m.weight = torch.nn.Parameter(torch.tensor(weight_up, dtype=torch.float32))
>     # torch_m.to("cuda")
>     torch_bn = torch.nn.BatchNorm2d(3).cuda()
>     torch_y = torch_bn(torch_imgs)
>     return torch_y.cpu().detach().numpy()
> 
> def oneflow_cal(np_arr: np.number, weight_up = None)-> np.number:
>     import oneflow as torch
>     torch_imgs = torch.FloatTensor(np_arr).cuda()
>     # print(torch_imgs.dtype)
>     # torch_m = torch.nn.Conv2d(in_channels, out_channels, 6, 6, 2, 2, bias=False)
>     # torch_m.weight = torch.nn.Parameter(torch.tensor(weight_up, dtype=torch.float32))
>     # torch_m.to("cuda")
>     torch_bn = torch.nn.BatchNorm2d(3).cuda()
>     torch_y = torch_bn(torch_imgs)
>     return torch_y.cpu().detach().numpy()
> 
> 
> def test_01(len : int):
>     for i in range(10):
>         np_arr = np.random.random(input_arr)
>         a = torch_cal(np_arr)
>         b = oneflow_cal(np_arr)
>         res = np.allclose(a, b)
>         print(res)
>         if res == False:
>             print(a.flatten().tolist()[0:15])
>             print(b.flatten().tolist()[0:15])
>             break
>            
> 
> 
> 
> # 随机生成 20 numpy测试 
> test_01(20)
> ```
> 
> ```
> False
> [1.6659996509552002, -0.7920742034912109, 1.3852777481079102, -0.3436308205127716, 1.0656812191009521, -1.3124674558639526, -1.5769178867340088, 1.3032417297363281, -0.16378989815711975, 0.3524649739265442, -0.8447245359420776, 1.1981629133224487, -1.5700066089630127, 0.4203341007232666, 0.928550124168396]
> [1.6659995317459106, -0.7920739650726318, 1.3852777481079102, -0.34363067150115967, 1.0656812191009521, -1.3124672174453735, -1.5769174098968506, 1.3032417297363281, -0.163789764046669, 0.35246503353118896, -0.8447242975234985, 1.1981627941131592, -1.570006251335144, 0.420334130525589, 0.928550124168396]
> ```



> -0.7920742

这里的结果仍然不能说明问题，1e-5的精度可以理解为是浮点数累加顺序不同导致的误差，而在yololv5中可以证明一个bn层导致的误差来到了1e-3的级别，需要去拿一下yolov5中的最小复现样例，复现出1e-3的差距。

**Comment by BBuf** [comment]
上周我们在one-yolo上调试，发现的问题为BN导致了1e-4级别的误差并且随着网络的推理这个误差被逐级加深，导致整个网络loss曲线偏差很大，并训练失败。如果其它同事关于这一点有经验可以和我们交流一下，我们稍后提供一个能反应1e-4精度误差的BN2d最小复现样例。

**Comment by Ldpe2G** [comment]
> 上周我们在one-yolo上调试，发现的问题为BN导致了1e-4级别的误差并且随着网络的推理这个误差被逐级放大，导致整个网络loss曲线偏差很大，并训练失败。如果其它同事关于这一点有经验可以和我们交流一下，我们稍后提供一个能反应1e-4精度误差的BN2d最小复现样例。

是gpu上运行bn有误差还是cpu，之前有个相关的Issue 
https://github.com/Oneflow-Inc/oneflow/issues/6526#issuecomment-962975775

**Comment by BBuf** [comment]
> > 上周我们在one-yolo上调试，发现的问题为BN导致了1e-4级别的误差并且随着网络的推理这个误差被逐级放大，导致整个网络loss曲线偏差很大，并训练失败。如果其它同事关于这一点有经验可以和我们交流一下，我们稍后提供一个能反应1e-4精度误差的BN2d最小复现样例。
> 
> 是gpu上运行bn有误差还是cpu，之前有个相关的Issue [#6526 (comment)](https://github.com/Oneflow-Inc/oneflow/issues/6526#issuecomment-962975775)

肯定是gpu，整理好复现样例之后我们约个会看下。

**Comment by Ldpe2G** [comment]
经过实验发现，oneflow 的 batchnorm 虽然和 torch 的batchnorm 都是调用了 cudnn 的实现，但是对于 `mode` 这个参数 oneflow 固定是 `CUDNN_BATCHNORM_SPATIAL_PERSISTENT`。

而对于 torch 在输入格式是 NCHW 的时候选用的是 `CUDNN_BATCHNORM_SPATIAL`
https://github.com/pytorch/pytorch/blob/e926f75b0bc66c789365cb1c48ba41e8447b97fb/aten/src/ATen/native/cudnn/BatchNorm.cpp#L71

torch 里加了一段注释：
```C++
    // TODO: The new CUDNN_BATCHNORM_SPATIAL_PERSISTENT mode was
    // introduced in CuDNN 7 for performance optimization, but it results in
    // accuracy losses in convolution models such as ResNeXt-101 and
    // video R(2+1)D. We will fall back to the normal CUDNN_BATCHNORM_SPATIAL
```

然后将oneflow 里的 batchnorm 也改成 `CUDNN_BATCHNORM_SPATIAL` 之后和torch就能对齐了。

**Comment by BBuf** [comment]
> 

这个问题以前没人发现么？这个应该很容易让oneflow训练模型的时候掉点吧。:joy:

**Comment by yuanms2** [comment]
是只有NCHW 改用 CUDNN_BATCHNORM_SPATIAL, 还是所有情况都改回去？

改回去后，不知道oneflow在 cnn 之类的模型上精确率会不会超过torch，之前应该是花了很大功夫把精度对齐到一个水平了

**Comment by yuanms2** [comment]
![image](https://user-images.githubusercontent.com/4755689/184785061-90c08253-684b-48f1-be5d-258fa8267593.png)

**Comment by Ldpe2G** [comment]
> 是只有NCHW 改用 CUDNN_BATCHNORM_SPATIAL, 还是所有情况都改回去？
> 
> 改回去后，不知道oneflow在 cnn 之类的模型上精确率会不会超过torch，之前应该是花了很大功夫把精度对齐到一个水平了

yolo那个项目用到的是 NCHW 格式，晓雨给了一个真实数据的测试例子上测试发现改成 `CUDNN_BATCHNORM_SPATIAL` 才能和torch的bn输出对齐，我先粗略改了一个pr，让冯文测一下看看是不是因为这个导致不收敛。后面或者是参考torch的改法，或者加个格式判断， 不同格式采用不同的mode

**Comment by Ldpe2G** [comment]
paddle 在 19 年的时候提了个 类似的 pr 默认不用 PERSISTENT 模式

https://github.com/PaddlePaddle/Paddle/pull/16545

**Comment by yuanms2** [comment]
Paddle 有个flag，可以配置是否使用PERSISTENT模式，默认不用

**Comment by Ldpe2G** [comment]
对齐之后的实验结果：
https://github.com/Oneflow-Inc/one-yolo/issues/10
